### DISCUSSION TITLE
Best practice to use React Components in Component Blocks

### DISCUSSION THREAD

**Discussion by andre-lergier** [keystonejs/keystone#7416] [Questions] [answered]
I’m new to Keystone and currently trying out some stuff.
One of the most important features to me are **Component Blocks**. For me the [documentation](https://keystonejs.com/docs/guides/document-fields#component-blocks) to this topic isn’t really clear so I would be really thankful for your best practice knowledge.

I have the following code for a simple Component Block:
```js
import React from 'react';
import { NotEditable, component, fields } from '@keystone-6/fields-document/component-blocks';

export const componentBlocks = {
  test: component({
    component: ({ text }) => {
      return (
        <div>
          <NotEditable>{text.value}</NotEditable>
        </div>
      );
    },
    label: 'Test',
    props: {
      text: fields.text({ label: 'Label...', defaultValue: 'defaultText' }),
    },
    chromeless: false,
  }),
}
```

**Now my first question:**
When do I need to use the `<NotEditable>Some Text</NotEditable> tag`? I can’t spot a difference if I put the tag or not.

**My second question:**
What’s your best practice with your React Components? Do you use/import your final components inside the `component-blocks.tsx` file? So your final components contain content like the `<NotEditable>` tag? Or do you create preview versions of the components which are used inside the document field in Keystone and thus are separated from the final frontend components?

Thanks for sharing your experience and knowledge!

**Comment by dcousens** [comment, marked as answer]
These are good questions @andre-lergier!
I think both of these questions can be answered with some context, and maybe the [document fields guide](https://keystonejs.com/docs/guides/document-fields#child-fields) could use a renovation.

#### Useful context
The document editor (`@keystone-6/fields-document`) is a rich text editor, where children components are rendered within a `contentEditable` root node.  This results in subsequent children nodes inheriting `contentEditable`,  and thus being editable.

Unfortunately,  when you're writing React components, this isn't exactly what we're used to; and we are put in a position where a `<span>hello world</span>` could be edited and changed (with no useful outcome for your component preview, and possibly an error).

### 1. When do you use `NotEditable`
If you don't want your `hello world` to have editable text, then in that scenario you should wrap your component with `NotEditable`.
`NotEditable` wraps it's children and changes this behaviour.
https://github.com/keystonejs/keystone/blob/4a75e3423830a916573cc45b5acc6b4d5fb65e01/packages/fields-document/src/DocumentEditor/component-blocks/api.tsx#L568-L569

### 2. What’s your best practice with React Components?
A first-pass approach for moving your frontend React components into components that can be edited within the document editor, could be wrapping them completely in a root node `NotEditable`,  thereby rendering them as-is.

This first-pass provides a non-inline editable experience where you would use the edit form to modify the relevant field values that end up being provided as props to your preview component.
This is often 99% of functionality you might actually want; and results in a prop schema consisting of [form fields, object fields and relationship fields](https://keystonejs.com/docs/guides/document-fields#form-fields).

If however, you want to increase the fidelity with which your users interact with the frontend components to _not_ be limited to the edit form,  this is where your frontend components and preview components will fork and you will use [child fields](https://keystonejs.com/docs/guides/document-fields#child-fields).
The extent to which they fork may depend on their complexity and your capabilities to inject these editable sub-components into the relevant components; but you will find yourself using `NotEditable` sprinkled aplenty.
